interfaces/builtin/jack.go

```diff
@@ -0,0 +1,52 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+const jackSummary = `allows access to the jack socket`
+
+const jackBaseDeclarationSlots = `
+  jack:
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: true
+`
+
+/*
+ * libjack builds the shared memory endpoint like this :
+ * jack-<userid>/<server name>/jack-<server id>
+ *
+ * see libjack/shm.c in jack1's source tree,
+ * and common/shm.c in jack2's source tree.
+ */
+const jackConnectedPlugAppArmor = `
+owner /dev/shm/jack-[0-9]*/*/* rw,
```

Comment:
I am not familiar with jack, so I googled and came across https://help.ubuntu.com/community/UbuntuStudioPreparation#Jack_-_jackd which suggests installing qjackctl, so in and Ubuntu 18.04 Desktop VM, I did, launched it and pressed 'start' and found the following in /dev/shm:
```
jack-1000-0          jack_sem.1000_default_dbusapi    jack_sem.1000_default_system
jack-1000-1          jack_sem.1000_default_freewheel  jack-shm-registry
jack_default_1000_0  jack_sem.1000_default_qjackctl
```

As you can see, none of these paths works with the specified rule and it isn't clear to me which of these files the snap should have access to. Additionally, jack-client-test couldn't find the files in the first place:
```
connect(2) call to /dev/shm/jack-1000/default/jack_0 failed (err=No such file or directory)
attempt to connect to server failed
jack_client_open() failed, status = 0x11
Unable to connect to JACK server
```

Reply:
This makes it more clear what all those files are:
```
$ ls -l /dev/shm
total 23060
-rw-r--r-- 1 jamie jamie 82280346 Apr 11 17:00 jack-1000-0
-rw-r--r-- 1 jamie jamie     1187 Apr 11 17:00 jack-1000-1
srwxr-xr-x 1 jamie jamie        0 Apr 11 17:00 jack_default_1000_0
-rwxr-xr-x 1 jamie jamie       12 Apr 11 17:00 jack_sem.1000_default_dbusapi
-rwxr-xr-x 1 jamie jamie       12 Apr 11 17:00 jack_sem.1000_default_freewheel
-rwxr-xr-x 1 jamie jamie       12 Apr 11 17:00 jack_sem.1000_default_qjackctl
-rwxr-xr-x 1 jamie jamie       12 Apr 11 17:00 jack_sem.1000_default_system
-rw-rw-rw- 1 jamie jamie    70744 Apr 11 16:47 jack-shm-registry
```

Reply:
Interesting. I don't have that at all, even though I'm running jack through QJackCtl too :
```
yannick@heimdall:~$ ls -l /dev/shm/
total 2124
drwx------ 3 yannick yannick      60 avr 12 00:07 jack-1000
-rwxrwxr-x 1 yannick yannick    1040 avr  8 22:11 u1000-Shm_125376fc
-rwxrwxr-x 1 yannick yannick   32784 avr  8 22:11 u1000-Shm_61df3bf5
-rwxrwxr-x 1 yannick yannick   32784 avr  8 22:11 u1000-Shm_78b2a20b
-rwxrwxr-x 1 yannick yannick 2097152 avr  8 22:11 u1000-ValveIPCSharedObj-Steam
```
Only `jack-1000` is relevant here, and matches what can be found in `libjack`'s source code.

Reply:
Oh ! I bet you are using jack2. I'm using jack1. 
The naming of the shm endpoint apparently changed between those 2 versions.
jack1 : `libjack/shm.c`, function `jack_shmalloc` : 
```c
snprintf (name, sizeof(name), "/jack-%d", registry->index);
```

jack2 : `common/shm.c`, function `jack_shmalloc` : 
```c
snprintf (name, sizeof (name), "jack-%d-%d", GetUID(), registry->index);
```

So I guess this interface is only for jack1.

Reply:
Yep, confirmed : here's what I have when running jack2 :
```
$ ls -l /dev/shm
-rw-r--r-- 1 yannick yannick 82280346 avr 12 00:32 jack-1000-0
-rw-r--r-- 1 yannick yannick     1187 avr 12 00:32 jack-1000-1
srwxr-xr-x 1 yannick yannick        0 avr 12 00:32 jack_default_1000_0
-rwxr-xr-x 1 yannick yannick       12 avr 12 00:32 jack_sem.1000_default_freewheel
-rwxr-xr-x 1 yannick yannick       12 avr 12 00:32 jack_sem.1000_default_system
-rw-rw-rw- 1 yannick yannick    70744 avr 12 00:32 jack-shm-registry
....
```

FYI, I've tested the following applications :
- iDJC (streamer/recorder)
- Ardour5 (DAW)
- Catia (patchbay)

None of them work with jack2. They work wonderfully with jack1

Reply:
I've been wanting to look at this and see if we could support both jack1 and jack2 in the same interface. While I've not been able to do this yet, I just wanted to let others know what I was thinking. Hopefully I can get to this soon (else, @ymauray, feel free to push on jack2 in the meantime if you're so inclined).

Reply:
Interesting, I guess I could try and see if I can make that work for both jack 1 and jack 2.
I was, however, under the impression that jack2 was not as popular as jack1.
It is definitely not compatible.
Therefore, is it wise to have one interface for 2 essentially different things ?

Reply:
This is a good point especially when considering a future slot implementation that provides the jack server slot for consumers (plugs). From a pure policy POV for an implicit classic interface where we don't know what the server is (since it is provided as a deb), I was thinking it was ok to treat it as more of an apparmor abstraction. However interfaces are not just collections of rules and interfaces were designed with the idea that different api-breaking slot implementations would need different interface since interfaces form a contract between the consumer and the provider. In this case, if we did as I suggested (where in the future we implement the non-implicit interface code) then a user might install a jack2 provider and a jack1 consumer, use 'snap connect' to connect them and things are broken.

As such, I'll review this PR with only jack1 in mind, and if you or someone else is so inclined, afterward someone can submit a jack2 PR.

Reply:
> I am not familiar with jack, so I googled and came across https://help.ubuntu.com/community/UbuntuStudioPreparation#Jack_-_jackd which suggests installing qjackctl, so in an Ubuntu 18.04 Desktop VM, I did, launched it and pressed 'start' and found the following in /dev/shm:

To verify JACK1, I installed jackd1, launched qjackctl, went into Settings, unchecked 'realtime', then pressed 'Start' and found the following in /dev/shm:
```
$ ls -lR jack-1000/
jack-1000/:
total 0
drwx------ 2 jamie jamie 140 May 24 08:44 default

jack-1000/default:
total 0
srwxr-xr-x 1 jamie jamie 0 May 24 08:44 jack_0
srwxr-xr-x 1 jamie jamie 0 May 24 08:44 jack_ack_0
prw-r--r-- 1 jamie jamie 0 May 24 08:44 jack-ack-fifo-8641-0
prw-r--r-- 1 jamie jamie 0 May 24 08:44 jack-ack-fifo-8641-1
prw-r--r-- 1 jamie jamie 0 May 24 08:44 jack-ack-fifo-8641-2
```

I then installed jack-client-test, which has no plugs, and added the above apparmor rules and was able to run the client. Therefore, all required seccomp rules are in the default template.